cmd/pulumi-test-language/interface.go

```diff
@@ -0,0 +1,861 @@
+	return &testingrpc.RunLanguageTestResponse{
+		Success:  false,
+		Messages: []string{msg},
+	}
+}
+
+func copyDirectory(fs iofs.FS, src string, dst string) error {
+	return iofs.WalkDir(fs, src, func(path string, d iofs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		relativePath, err := filepath.Rel(src, path)
+		contract.AssertNoErrorf(err, "path %s should be relative to %s", path, src)
+
+		srcPath := filepath.Join(src, relativePath)
+		dstPath := filepath.Join(dst, relativePath)
+		contract.Assertf(srcPath == path, "srcPath %s should be equal to path %s", srcPath, path)
+
+		if d.IsDir() {
+			return os.MkdirAll(dstPath, 0o700)
+		}
+
+		srcFile, err := fs.Open(srcPath)
+		if err != nil {
+			return fmt.Errorf("open file %s: %w", srcPath, err)
+		}
+		defer srcFile.Close()
+
+		dstFile, err := os.Create(dstPath)
+		if err != nil {
+			return fmt.Errorf("create file %s: %w", dstPath, err)
+		}
+		defer dstFile.Close()
+
+		_, err = io.Copy(dstFile, srcFile)
+		if err != nil {
+			return fmt.Errorf("copy file %s->%s: %w", srcPath, dstPath, err)
+		}
+
+		return nil
+	})
+}
+
+// compareDirectories compares two directories, returning a list of validation failures where files had
+// different contents, or we only present in on of the directories. If allowNewFiles is true then it's ok to
+// have extra files in the actual directory, we use this for checking building SDKs doesn't mutate any files,
+// but doing so might add new build files (which would then normally be .gitignored).
+func compareDirectories(actualDir, expectedDir string, allowNewFiles bool) ([]string, error) {
+	var validations []string
+	// Check that every file in expected is also in actual with the same content
+	err := filepath.WalkDir(expectedDir, func(path string, d iofs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		// No need to check directories, just recurse into them
+		if d.IsDir() {
+			return nil
+		}
+
+		relativePath, err := filepath.Rel(expectedDir, path)
+		contract.AssertNoErrorf(err, "path %s should be relative to %s", path, expectedDir)
+
+		// Check that the file is present in the expected directory and has the same contents
+		expectedContents, err := os.ReadFile(filepath.Join(expectedDir, relativePath))
+		if err != nil {
+			return fmt.Errorf("read expected file: %w", err)
+		}
+
+		actualPath := filepath.Join(actualDir, relativePath)
+		actualContents, err := os.ReadFile(actualPath)
+		// An error here is a test failure rather than an error, add this to the validation list
+		if err != nil {
+			validations = append(validations, fmt.Sprintf("expected file %s could not be read", relativePath))
+			// Move on to the next file
+			return nil
+		}
+
+		if !bytes.Equal(actualContents, expectedContents) {
+			// TODO: Find a way to show better diffs here
+			validations = append(validations, fmt.Sprintf("expected file %s does not match actual file", relativePath))
```

Comment:
Consider [go-cmp](https://pkg.go.dev/github.com/google/go-cmp/cmp) or [pkg/diff](https://pkg.go.dev/github.com/pkg/diff). The latter prints something closer to a `git diff`-style unified diff.

Reply:
Good idea